{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Analysis\n",
    "\n",
    "In this file, we proceed to develop and test the pipeline that handles the \"trancripts\" dataset.\n",
    "\n",
    "We then proceed to analyse the output of said pipeline later on"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All the imports for all the things"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import datetime\n",
    "\n",
    "\n",
    "import pathlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "import importlib\n",
    "\n",
    "\n",
    "from tensorflow.keras import layers, losses\n",
    "from tensorflow.keras.models import Model\n",
    "from tensorflow import keras\n",
    "from tensorflow.keras import callbacks  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from tensorflow.python.ops.numpy_ops import np_config\n",
    "np_config.enable_numpy_behavior()\n",
    "\n",
    "import plotly.express as px\n",
    "import plotly.subplots as sp\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "# for translation of gene symbols\n",
    "import mygene\n",
    "\n",
    "import pathlib\n",
    "import scipy\n",
    "\n",
    "from sklearn.preprocessing import normalize, MinMaxScaler\n",
    "\n",
    "# to import libs \n",
    "sys.path.append('../src/utils')\n",
    "import  feature_selection, visualisation\n",
    "\n",
    "sys.path.append('../src')\n",
    "import data_handler\n",
    "\n",
    "\n",
    "\n",
    "# datasets\n",
    "\n",
    "# default path of the folder containing the salmon files\n",
    "absolute_path = '/Users/aygalic/Thesis/data/quant/'  \n",
    "metadata_path = '/Users/aygalic/Thesis/METADATA_200123.xlsx'  \n",
    "\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "pd.options.display.width = 1000\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some testing about gene name retrieval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PPMI-Phase2-IR2.41282.V02.0003241603.5104-SL-4884.longRNA-NEBKAP.salmon-gtf.transcripts.sf'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# each couple of entries correspond to one patient, we are only interested in the \"transcript\" files\n",
    "entries = os.listdir(absolute_path)\n",
    "#entries_transcripts = [e for e in entries if \"transcripts\" in e ]\n",
    "entries = [e for e in entries if \"transcripts\" in e ]\n",
    "entry = entries[0]\n",
    "entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "95309\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...\n",
       "1    ENST00000423372.3|ENSG00000237683.5|-|-|AL6273...\n",
       "2    ENST00000426406.1|ENSG00000235249.1|OTTHUMG000...\n",
       "Name: Name, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = data_handler.get_names(os.path.join(absolute_path, entry))\n",
    "print(len(gene_names))\n",
    "gene_names[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        trascript_id            gene_id                   idk                   idk transcript_variant      symbol length untranslated_region_3 coding_region untranslated_region_5   idk\n",
      "0  ENST00000335137.3  ENSG00000186092.4  OTTHUMG00000001094.1  OTTHUMT00000003223.1          OR4F5-001       OR4F5    918             CDS:1-918                                None  None\n",
      "1  ENST00000423372.3  ENSG00000237683.5                     -                     -     AL627309.1-201  AL627309.1   2661             UTR5:1-70    CDS:71-850         UTR3:851-2661      \n",
      "2  ENST00000426406.1  ENSG00000235249.1  OTTHUMG00000002860.1  OTTHUMT00000007999.1         OR4F29-001      OR4F29    995             UTR5:1-19    CDS:20-958          UTR3:959-995      \n",
      "3  ENST00000332831.2  ENSG00000185097.2  OTTHUMG00000002581.1  OTTHUMT00000007334.1         OR4F16-001      OR4F16    995             UTR5:1-19    CDS:20-958          UTR3:959-995      \n",
      "4  ENST00000599533.1  ENSG00000269831.1                     -                     -     AL669831.1-201  AL669831.1    129             CDS:1-129                                None  None\n"
     ]
    }
   ],
   "source": [
    "name_df = pd.DataFrame([gn.split(\"|\") for gn in gene_names])\n",
    "name_df = name_df.set_axis([\n",
    "    'trascript_id', \n",
    "    'gene_id', \n",
    "    'idk', \n",
    "    'idk', \n",
    "    'transcript_variant', \n",
    "    'symbol', \n",
    "    'length', \n",
    "    'untranslated_region_3', \n",
    "    'coding_region', \n",
    "    'untranslated_region_5', \n",
    "    'idk'], axis=1)\n",
    "\n",
    "print(name_df[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0         OR4F5\n",
      "1    AL627309.1\n",
      "2        OR4F29\n",
      "3        OR4F16\n",
      "4    AL669831.1\n",
      "Name: symbol, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(name_df[\"symbol\"][0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# giving a chance to supervisor provided genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           gene_id          baseMean    log2FoldChange              lfcSE              stat               pvalue                padj      name\n",
      "1  ENSG00000269981  10,5643592923569  1,71531890942766  0,579755024617464  2,95869606401336  0,00308943642200947  0,0417143018059277       NaN\n",
      "2  ENSG00000279928  13,3318505597289  1,09761280588101   0,38628946637286  2,84142567020337  0,00449123202503173  0,0477316305084926  DDX11L17\n",
      "3  ENSG00000177757  4,71120273203759  2,45838296821385  0,772903430141131  3,18071168058467  0,00146913763891777  0,0336358229549266    FAM87B\n",
      "4  ENSG00000230699  6,40709033425367  1,82833056439525  0,650787408589311  2,80941293618212  0,00496319413781045  0,0493394233571598       NaN\n"
     ]
    }
   ],
   "source": [
    "suggested_genes_files = '../../data/PD_genes_of_interest/signif_de_Disease_status_IDIOPATHIC_PD_01.tsv'\n",
    "suggested_genes = pd.read_csv(suggested_genes_files, sep='\\t')\n",
    "suggested_genes = suggested_genes.rename(columns={'Unnamed: 0': 'gene_id'})\n",
    "print(suggested_genes[1:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol\n",
       "GPR56    70\n",
       "NDRG2    61\n",
       "EEF1D    49\n",
       "DMKN     49\n",
       "DDR1     47\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_df[\"symbol\"].value_counts().head(5) # Print the top 5 occurrence\n",
    "# assumption : NaN would show up if they were any..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "ENSG00000205336.7     70\n",
       "ENSG00000165795.16    61\n",
       "ENSG00000104529.13    49\n",
       "ENSG00000161249.16    49\n",
       "ENSG00000204580.7     47\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_df[\"gene_id\"].value_counts().head(5) # Print the top 15 occurrence\n",
    "# same results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8395\n",
      "8988\n",
      "9061\n"
     ]
    }
   ],
   "source": [
    "# masking approach\n",
    "mask_gene_name = name_df[\"symbol\"].dropna().isin(suggested_genes[\"name\"]) # there doesn't appear to be any NA but fine\n",
    "#mask_gene_name = name_df[\"symbol\"].isin(suggested_genes[\"name\"])\n",
    "print(sum(mask_gene_name))\n",
    "\n",
    "mask_gene_id = pd.Series([id.split(\".\")[0] for id in name_df[\"gene_id\"]]).isin(suggested_genes[\"gene_id\"])\n",
    "print(sum(mask_gene_id))\n",
    "\n",
    "\n",
    "combined_mask = mask_gene_name | mask_gene_id\n",
    "print(sum(combined_mask))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Just a bit of testing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retaining patients that are included in phases 1 & 2\n",
      "retaining all patient who have at least passed the Base Line Visit...\n",
      "loading samples...\n",
      "loaded 2853 samples\n",
      "number of genes selected :  9061\n",
      "log(1 + x) transformation...\n",
      "scaling to [0, 1]...\n",
      "number of seq in the dataset : 2853\n",
      "keeping sample as is, no conversion to time series\n"
     ]
    }
   ],
   "source": [
    "importlib.reload(feature_selection) # to allow modification of the script without restarting the whole session\n",
    "importlib.reload(data_handler) # to allow modification of the script without restarting the whole session\n",
    "\n",
    "\n",
    "as_time_series = False\n",
    "dataset, index_names, n_genes, gene_names = data_handler.generate_dataset_transcripts(feature_selection_threshold = None, \n",
    "                                                            batch_size = 32, \n",
    "                                                            retain_phases = \"Both\",\n",
    "                                                            normalization = False,\n",
    "                                                            as_time_series = as_time_series,\n",
    "                                                            MT_removal = False,\n",
    "                                                            gene_selection_file = suggested_genes_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "get the dataset as a numpy object for data manipulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get everything out of TensorFlow back to numpy/pandas\n",
    "data = np.concatenate(list(dataset.as_numpy_iterator()), axis=0)\n",
    "if(as_time_series):\n",
    "    data = data.reshape(data.shape[0]*data.shape[1], -1)\n",
    "    data = pd.DataFrame(data)\n",
    "    data.columns = gene_names[\"name\"]\n",
    "    data.index = np.repeat(index_names, 5, axis=0)\n",
    "else:\n",
    "    data = pd.DataFrame(data)\n",
    "    data.columns = gene_names[\"name\"]\n",
    "    data.index = index_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name      TOP1  ENSGR0000168939  ENSGR0000214717  ENSGR0000182484   \n",
      "3442  0.314952         0.360238         0.136580         0.517728  \\\n",
      "3442  0.314498         0.420334         0.281522         0.490566   \n",
      "3442  0.742467         0.354860         0.261963         0.379775   \n",
      "3442  0.563241         0.467921         0.193135         0.520309   \n",
      "3442  0.344854         0.334907         0.132422         0.420329   \n",
      "\n",
      "name  ENSGR0000198223  ENSGR0000185960  ENSGR0000205755  ENSGR0000185291   \n",
      "3442        -0.384073        -0.030902         0.462116         0.471284  \\\n",
      "3442        -0.136599         0.229172         0.495155         0.485320   \n",
      "3442        -0.313197         0.386343         0.304953         0.417484   \n",
      "3442        -0.159063        -0.018419         0.390361         0.396865   \n",
      "3442        -0.456131        -0.360034         0.325078         0.271394   \n",
      "\n",
      "name  ENSGR0000002586     KIF5A  ...    GARRE1    PTMAP2    TSEN34     MAPK4   \n",
      "3442        -0.187010  0.360238  ...  0.179150  0.598161  0.440379  0.069291  \\\n",
      "3442        -0.177320  0.420334  ... -0.021836  0.585829  0.353560  0.355534   \n",
      "3442         0.112342  0.354860  ...  0.068308  0.186448 -1.000000 -0.369964   \n",
      "3442        -0.091810  0.467921  ...  0.126058  0.322702  0.126988 -0.204413   \n",
      "3442        -0.189894  0.334907  ... -0.196816  0.448174  0.259950  0.066665   \n",
      "\n",
      "name   PHYHIPL     KANK2  ENSG00000269588  PARD6G-AS1     GCNT1      CBX2  \n",
      "3442  0.059727  0.158673        -0.089087    0.750658  0.118729 -0.230801  \n",
      "3442  0.102290  0.224923        -0.375323    0.552260  0.358060 -0.059470  \n",
      "3442  0.375380  0.436012         0.208570    0.560815  0.341015  0.135144  \n",
      "3442 -0.047807  0.112144         0.328376    0.778394 -0.051124 -0.035061  \n",
      "3442  0.088396 -0.008932         0.068408    0.655429  0.054654 -0.205888  \n",
      "\n",
      "[5 rows x 11957 columns]\n"
     ]
    }
   ],
   "source": [
    "print(data.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retaining patients that are included in phases 1 & 2\n",
      "retaining all patient who have at least passed the Base Line Visit...\n",
      "loading samples...\n",
      "loaded 27 samples\n",
      "retriving symbols for genes\n",
      "number of seq in the dataset : 27\n",
      "keeping sample as is, no conversion to time series\n"
     ]
    }
   ],
   "source": [
    "importlib.reload(feature_selection) # to allow modification of the script without restarting the whole session\n",
    "\n",
    "as_time_series = False\n",
    "\n",
    "dataset, index_names, n_genes, gene_names = generate_dataset_transcripts(\n",
    "    feature_selection_threshold = 2, \n",
    "    subsample= 100,\n",
    "    retain_phases = \"Both\",\n",
    "    normalization = False,\n",
    "    as_time_series = as_time_series,\n",
    "    MT_removal = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        False\n",
       "1        False\n",
       "2        False\n",
       "3        False\n",
       "4        False\n",
       "         ...  \n",
       "95304     True\n",
       "95305     True\n",
       "95306     True\n",
       "95307     True\n",
       "95308     True\n",
       "Length: 95309, dtype: bool"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([n.split(\"-\")[0] for n in gene_names[4]]).duplicated()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "74674\n",
      "153\n"
     ]
    }
   ],
   "source": [
    "# absurd amount of duplicates\n",
    "print(sum(pd.DataFrame([s.rsplit('-', 1)[0] for s in gene_names[4]]).duplicated()))\n",
    "print(sum(pd.DataFrame([s for s in gene_names[4]]).duplicated()))\n",
    "\n",
    "\n",
    "index = 0\n",
    "print(pd.DataFrame([s for s in gene_names[index]]))\n",
    "print(sum(pd.DataFrame([s for s in gene_names[index]]).duplicated()))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       0\n",
      "0      ENST00000335137.3\n",
      "1      ENST00000423372.3\n",
      "2      ENST00000426406.1\n",
      "3      ENST00000332831.2\n",
      "4      ENST00000599533.1\n",
      "...                  ...\n",
      "95304  ENST00000361335.1\n",
      "95305  ENST00000361381.2\n",
      "95306  ENST00000361567.2\n",
      "95307  ENST00000361681.2\n",
      "95308  ENST00000361789.2\n",
      "\n",
      "[95309 rows x 1 columns]\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "index = 0\n",
    "print(pd.DataFrame([s for s in gene_names[index]]))\n",
    "print(sum(pd.DataFrame([s for s in gene_names[index]]).duplicated()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['OR4F5',\n",
       " 'AL627309',\n",
       " 'OR4F29',\n",
       " 'OR4F16',\n",
       " 'AL669831',\n",
       " 'AL645608',\n",
       " 'SAMD11',\n",
       " 'SAMD11',\n",
       " 'SAMD11',\n",
       " 'SAMD11',\n",
       " 'SAMD11',\n",
       " 'AL645608',\n",
       " 'NOC2L',\n",
       " 'KLHL17',\n",
       " 'KLHL17',\n",
       " 'PLEKHN1',\n",
       " 'PLEKHN1',\n",
       " 'PLEKHN1',\n",
       " 'PLEKHN1',\n",
       " 'C1orf170',\n",
       " 'C1orf170',\n",
       " 'HES4',\n",
       " 'HES4',\n",
       " 'HES4',\n",
       " 'ISG15',\n",
       " 'AGRN',\n",
       " 'AGRN',\n",
       " 'RNF223',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'C1orf159',\n",
       " 'TTLL10',\n",
       " 'TTLL10',\n",
       " 'TTLL10',\n",
       " 'TTLL10',\n",
       " 'TNFRSF18',\n",
       " 'TNFRSF18',\n",
       " 'TNFRSF18',\n",
       " 'TNFRSF18',\n",
       " 'TNFRSF4',\n",
       " 'SDF4',\n",
       " 'SDF4',\n",
       " 'SDF4',\n",
       " 'SDF4',\n",
       " 'SDF4',\n",
       " 'B3GALT6',\n",
       " 'FAM132A',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'UBE2J2',\n",
       " 'SCNN1D',\n",
       " 'SCNN1D',\n",
       " 'SCNN1D',\n",
       " 'SCNN1D',\n",
       " 'SCNN1D',\n",
       " 'SCNN1D',\n",
       " 'SCNN1D',\n",
       " 'ACAP3',\n",
       " 'ACAP3',\n",
       " 'ACAP3',\n",
       " 'PUSL1',\n",
       " 'PUSL1',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'CPSF3L',\n",
       " 'GLTPD1',\n",
       " 'GLTPD1',\n",
       " 'TAS1R3',\n",
       " 'DVL1',\n",
       " 'DVL1',\n",
       " 'MXRA8',\n",
       " 'MXRA8',\n",
       " 'MXRA8',\n",
       " 'MXRA8',\n",
       " 'AURKAIP1',\n",
       " 'AURKAIP1',\n",
       " 'AURKAIP1',\n",
       " 'AURKAIP1',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'CCNL2',\n",
       " 'RP4',\n",
       " 'RP4',\n",
       " 'RP4',\n",
       " 'RP4',\n",
       " 'RP4',\n",
       " 'MRPL20',\n",
       " 'MRPL20',\n",
       " 'MRPL20',\n",
       " 'ANKRD65',\n",
       " 'ANKRD65',\n",
       " 'ANKRD65',\n",
       " 'ANKRD65',\n",
       " 'ANKRD65',\n",
       " 'TMEM88B',\n",
       " 'VWA1',\n",
       " 'VWA1',\n",
       " 'VWA1',\n",
       " 'VWA1',\n",
       " 'VWA1',\n",
       " 'ATAD3C',\n",
       " 'ATAD3C',\n",
       " 'ATAD3B',\n",
       " 'ATAD3B',\n",
       " 'ATAD3A',\n",
       " 'ATAD3A',\n",
       " 'ATAD3A',\n",
       " 'ATAD3A',\n",
       " 'ATAD3A',\n",
       " 'TMEM240',\n",
       " 'TMEM240',\n",
       " 'SSU72',\n",
       " 'SSU72',\n",
       " 'AL645728',\n",
       " 'C1orf233',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MIB2',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'MMP23B',\n",
       " 'CDK11B',\n",
       " 'CDK11B',\n",
       " 'CDK11B',\n",
       " 'CDK11B',\n",
       " 'CDK11B',\n",
       " 'CDK11B',\n",
       " 'SLC35E2B',\n",
       " 'SLC35E2B',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'CDK11A',\n",
       " 'SLC35E2',\n",
       " 'SLC35E2',\n",
       " 'SLC35E2',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'NADK',\n",
       " 'GNB1',\n",
       " 'GNB1',\n",
       " 'GNB1',\n",
       " 'GNB1',\n",
       " 'CALML6',\n",
       " 'CALML6',\n",
       " 'TMEM52',\n",
       " 'TMEM52',\n",
       " 'TMEM52',\n",
       " 'TMEM52',\n",
       " 'TMEM52',\n",
       " 'C1orf222',\n",
       " 'C1orf222',\n",
       " 'C1orf222',\n",
       " 'C1orf222',\n",
       " 'C1orf222',\n",
       " 'C1orf222',\n",
       " 'GABRD',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'PRKCZ',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'C1orf86',\n",
       " 'AL590822',\n",
       " 'AL590822',\n",
       " 'SKI',\n",
       " 'MORN1',\n",
       " 'MORN1',\n",
       " 'MORN1',\n",
       " 'MORN1',\n",
       " 'MORN1',\n",
       " 'AL589739',\n",
       " 'RER1',\n",
       " 'RER1',\n",
       " 'RER1',\n",
       " 'RER1',\n",
       " 'RER1',\n",
       " 'RER1',\n",
       " 'RER1',\n",
       " 'PEX10',\n",
       " 'PEX10',\n",
       " 'PEX10',\n",
       " 'PEX10',\n",
       " 'PEX10',\n",
       " 'PEX10',\n",
       " 'PEX10',\n",
       " 'PLCH2',\n",
       " 'PLCH2',\n",
       " 'PLCH2',\n",
       " 'PLCH2',\n",
       " 'PLCH2',\n",
       " 'PLCH2',\n",
       " 'PLCH2',\n",
       " 'PANK4',\n",
       " 'PANK4',\n",
       " 'PANK4',\n",
       " 'PANK4',\n",
       " 'PANK4',\n",
       " 'PANK4',\n",
       " 'PANK4',\n",
       " 'HES5',\n",
       " 'TNFRSF14',\n",
       " 'TNFRSF14',\n",
       " 'TNFRSF14',\n",
       " 'TNFRSF14',\n",
       " 'TNFRSF14',\n",
       " 'TNFRSF14',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'FAM213B',\n",
       " 'MMEL1',\n",
       " 'MMEL1',\n",
       " 'MMEL1',\n",
       " 'MMEL1',\n",
       " 'MMEL1',\n",
       " 'TTC34',\n",
       " 'TTC34',\n",
       " 'TTC34',\n",
       " 'ACTRT2',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'PRDM16',\n",
       " 'ARHGEF16',\n",
       " 'ARHGEF16',\n",
       " 'ARHGEF16',\n",
       " 'ARHGEF16',\n",
       " 'ARHGEF16',\n",
       " 'ARHGEF16',\n",
       " 'MEGF6',\n",
       " 'MEGF6',\n",
       " 'MEGF6',\n",
       " 'MEGF6',\n",
       " 'TPRG1L',\n",
       " 'TPRG1L',\n",
       " 'WRAP73',\n",
       " 'WRAP73',\n",
       " 'WRAP73',\n",
       " 'WRAP73',\n",
       " 'WRAP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'TP73',\n",
       " 'CCDC27',\n",
       " 'CCDC27',\n",
       " 'SMIM1',\n",
       " 'SMIM1',\n",
       " 'SMIM1',\n",
       " 'LRRC47',\n",
       " 'LRRC47',\n",
       " 'CEP104',\n",
       " 'CEP104',\n",
       " 'CEP104',\n",
       " 'CEP104',\n",
       " 'CEP104',\n",
       " 'CEP104',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'DFFB',\n",
       " 'C1orf174',\n",
       " 'AJAP1',\n",
       " 'AJAP1',\n",
       " 'NPHP4',\n",
       " 'NPHP4',\n",
       " 'NPHP4',\n",
       " 'NPHP4',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'KCNAB2',\n",
       " 'AL035406',\n",
       " 'CHD5',\n",
       " 'CHD5',\n",
       " 'CHD5',\n",
       " 'CHD5',\n",
       " 'CHD5',\n",
       " 'CHD5',\n",
       " 'CHD5',\n",
       " 'RPL22',\n",
       " 'RPL22',\n",
       " 'RPL22',\n",
       " 'RPL22',\n",
       " 'RPL22',\n",
       " 'RPL22',\n",
       " 'RPL22',\n",
       " 'RNF207',\n",
       " 'RNF207',\n",
       " 'ICMT',\n",
       " 'ICMT',\n",
       " 'ICMT',\n",
       " 'ICMT',\n",
       " 'HES3',\n",
       " 'GPR153',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'ACOT7',\n",
       " 'HES2',\n",
       " 'HES2',\n",
       " 'HES2',\n",
       " 'HES2',\n",
       " 'HES2',\n",
       " 'ESPN',\n",
       " 'ESPN',\n",
       " 'ESPN',\n",
       " 'ESPN',\n",
       " 'ESPN',\n",
       " 'ESPN',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'TNFRSF25',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'PLEKHG5',\n",
       " 'NOL9',\n",
       " 'TAS1R1',\n",
       " 'TAS1R1',\n",
       " 'TAS1R1',\n",
       " 'TAS1R1',\n",
       " 'TAS1R1',\n",
       " 'ZBTB48',\n",
       " 'ZBTB48',\n",
       " 'ZBTB48',\n",
       " 'ZBTB48',\n",
       " 'ZBTB48',\n",
       " 'KLHL21',\n",
       " 'KLHL21',\n",
       " 'KLHL21',\n",
       " 'KLHL21',\n",
       " 'KLHL21',\n",
       " 'PHF13',\n",
       " 'THAP3',\n",
       " 'THAP3',\n",
       " 'THAP3',\n",
       " 'THAP3',\n",
       " 'THAP3',\n",
       " 'DNAJC11',\n",
       " 'DNAJC11',\n",
       " 'DNAJC11',\n",
       " 'DNAJC11',\n",
       " 'DNAJC11',\n",
       " 'DNAJC11',\n",
       " 'DNAJC11',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'CAMTA1',\n",
       " 'VAMP3',\n",
       " 'VAMP3',\n",
       " 'PER3',\n",
       " 'PER3',\n",
       " 'PER3',\n",
       " 'UTS2',\n",
       " 'UTS2',\n",
       " 'UTS2',\n",
       " 'TNFRSF9',\n",
       " 'TNFRSF9',\n",
       " 'TNFRSF9',\n",
       " 'PARK7',\n",
       " 'PARK7',\n",
       " 'PARK7',\n",
       " 'PARK7',\n",
       " 'PARK7',\n",
       " 'PARK7',\n",
       " 'PARK7',\n",
       " 'ERRFI1',\n",
       " 'ERRFI1',\n",
       " 'ERRFI1',\n",
       " 'ERRFI1',\n",
       " 'ERRFI1',\n",
       " 'SLC45A1',\n",
       " 'SLC45A1',\n",
       " 'SLC45A1',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'RERE',\n",
       " 'ENO1',\n",
       " 'ENO1',\n",
       " 'ENO1',\n",
       " 'CA6',\n",
       " 'CA6',\n",
       " 'CA6',\n",
       " 'CA6',\n",
       " 'CA6',\n",
       " 'SLC2A7',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'SLC2A5',\n",
       " 'GPR157',\n",
       " 'GPR157',\n",
       " 'GPR157',\n",
       " 'H6PD',\n",
       " 'H6PD',\n",
       " 'SPSB1',\n",
       " 'SPSB1',\n",
       " 'SPSB1',\n",
       " 'SPSB1',\n",
       " 'SLC25A33',\n",
       " 'TMEM201',\n",
       " 'TMEM201',\n",
       " 'TMEM201',\n",
       " 'TMEM201',\n",
       " 'TMEM201',\n",
       " 'PIK3CD',\n",
       " 'PIK3CD',\n",
       " 'PIK3CD',\n",
       " 'PIK3CD',\n",
       " 'C1orf200',\n",
       " 'CLSTN1',\n",
       " 'CLSTN1',\n",
       " 'CLSTN1',\n",
       " 'CLSTN1',\n",
       " 'CTNNBIP1',\n",
       " 'CTNNBIP1',\n",
       " 'CTNNBIP1',\n",
       " 'CTNNBIP1',\n",
       " 'CTNNBIP1',\n",
       " 'LZIC',\n",
       " 'LZIC',\n",
       " 'LZIC',\n",
       " 'LZIC',\n",
       " 'LZIC',\n",
       " 'NMNAT1',\n",
       " 'NMNAT1',\n",
       " 'NMNAT1',\n",
       " 'NMNAT1',\n",
       " 'RBP7',\n",
       " 'RBP7',\n",
       " 'UBE4B',\n",
       " 'UBE4B',\n",
       " 'UBE4B',\n",
       " 'UBE4B',\n",
       " 'UBE4B',\n",
       " 'UBE4B',\n",
       " 'KIF1B',\n",
       " 'KIF1B',\n",
       " 'KIF1B',\n",
       " 'KIF1B',\n",
       " 'KIF1B',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'PGD',\n",
       " 'APITD1',\n",
       " 'APITD1',\n",
       " 'APITD1',\n",
       " 'APITD1',\n",
       " 'APITD1',\n",
       " 'APITD1',\n",
       " 'APITD1',\n",
       " 'CORT',\n",
       " 'CORT',\n",
       " 'DFFA',\n",
       " 'DFFA',\n",
       " 'DFFA',\n",
       " 'PEX14',\n",
       " 'PEX14',\n",
       " 'PEX14',\n",
       " 'CASZ1',\n",
       " 'CASZ1',\n",
       " 'CASZ1',\n",
       " 'C1orf127',\n",
       " 'C1orf127',\n",
       " 'C1orf127',\n",
       " 'C1orf127',\n",
       " 'C1orf127',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'TARDBP',\n",
       " 'MASP2',\n",
       " 'MASP2',\n",
       " 'SRM',\n",
       " 'SRM',\n",
       " 'SRM',\n",
       " 'SRM',\n",
       " 'SRM',\n",
       " 'EXOSC10',\n",
       " 'EXOSC10',\n",
       " 'EXOSC10',\n",
       " 'EXOSC10',\n",
       " 'MTOR',\n",
       " 'MTOR',\n",
       " 'MTOR',\n",
       " 'ANGPTL7',\n",
       " 'UBIAD1',\n",
       " 'UBIAD1',\n",
       " 'UBIAD1',\n",
       " 'UBIAD1',\n",
       " 'PTCHD2',\n",
       " 'PTCHD2',\n",
       " 'PTCHD2',\n",
       " 'FBXO2',\n",
       " 'FBXO2',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO44',\n",
       " 'FBXO6',\n",
       " 'FBXO6',\n",
       " 'MAD2L2',\n",
       " 'MAD2L2',\n",
       " 'MAD2L2',\n",
       " 'MAD2L2',\n",
       " 'MAD2L2',\n",
       " 'MAD2L2',\n",
       " 'MAD2L2',\n",
       " 'DRAXIN',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'AGTRAP',\n",
       " 'C1orf167',\n",
       " 'C1orf167',\n",
       " 'C1orf167',\n",
       " 'C1orf167',\n",
       " 'C1orf167',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'MTHFR',\n",
       " 'CLCN6',\n",
       " 'CLCN6',\n",
       " 'CLCN6',\n",
       " 'CLCN6',\n",
       " 'CLCN6',\n",
       " 'NPPA',\n",
       " 'NPPA',\n",
       " 'NPPB',\n",
       " 'KIAA2013',\n",
       " 'KIAA2013',\n",
       " 'PLOD1',\n",
       " 'PLOD1',\n",
       " 'PLOD1',\n",
       " 'PLOD1',\n",
       " 'MFN2',\n",
       " 'MFN2',\n",
       " 'MFN2',\n",
       " 'MIIP',\n",
       " 'MIIP',\n",
       " 'TNFRSF8',\n",
       " 'TNFRSF8',\n",
       " 'TNFRSF8',\n",
       " 'TNFRSF8',\n",
       " 'TNFRSF1B',\n",
       " 'TNFRSF1B',\n",
       " 'VPS13D',\n",
       " 'VPS13D',\n",
       " 'VPS13D',\n",
       " 'VPS13D',\n",
       " 'VPS13D',\n",
       " 'VPS13D',\n",
       " 'VPS13D',\n",
       " 'DHRS3',\n",
       " 'DHRS3',\n",
       " 'AADACL4',\n",
       " 'AADACL3',\n",
       " 'AADACL3',\n",
       " 'C1orf158',\n",
       " 'C1orf158',\n",
       " 'PRAMEF12',\n",
       " 'PRAMEF1',\n",
       " 'PRAMEF1',\n",
       " 'PRAMEF11',\n",
       " 'PRAMEF11',\n",
       " 'HNRNPCL1',\n",
       " 'PRAMEF2',\n",
       " 'PRAMEF4',\n",
       " 'PRAMEF10',\n",
       " 'PRAMEF7',\n",
       " 'PRAMEF7',\n",
       " 'PRAMEF6',\n",
       " 'PRAMEF6',\n",
       " 'PRAMEF6',\n",
       " 'PRAMEF6',\n",
       " 'PRAMEF6',\n",
       " 'PRAMEF6',\n",
       " 'PRAMEF22',\n",
       " 'WI2',\n",
       " 'PRAMEF26',\n",
       " 'PRAMEF3',\n",
       " 'PRAMEF3',\n",
       " 'PRAMEF5',\n",
       " 'PRAMEF8',\n",
       " 'PRAMEF9',\n",
       " 'PRAMEF13',\n",
       " 'PRAMEF18',\n",
       " 'PRAMEF16',\n",
       " 'PRAMEF21',\n",
       " 'PRAMEF21',\n",
       " 'PRAMEF15',\n",
       " 'PRAMEF14',\n",
       " 'PRAMEF14',\n",
       " 'PRAMEF19',\n",
       " 'PRAMEF19',\n",
       " 'PRAMEF17',\n",
       " 'PRAMEF20',\n",
       " 'PRAMEF20',\n",
       " 'LRRC38',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PDPN',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'PRDM2',\n",
       " 'KAZN',\n",
       " 'KAZN',\n",
       " 'KAZN',\n",
       " 'KAZN',\n",
       " 'KAZN',\n",
       " 'KAZN',\n",
       " 'KAZN',\n",
       " 'TMEM51',\n",
       " 'TMEM51',\n",
       " 'TMEM51',\n",
       " 'TMEM51',\n",
       " 'TMEM51',\n",
       " 'TMEM51',\n",
       " 'C1orf195',\n",
       " 'C1orf195',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'FHAD1',\n",
       " 'EFHD2',\n",
       " 'EFHD2',\n",
       " 'CTRC',\n",
       " 'CTRC',\n",
       " 'CELA2A',\n",
       " 'CELA2B',\n",
       " 'CELA2B',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'CASP9',\n",
       " 'DNAJC16',\n",
       " 'DNAJC16',\n",
       " 'DNAJC16',\n",
       " 'DNAJC16',\n",
       " 'DNAJC16',\n",
       " 'AGMAT',\n",
       " 'DDI2',\n",
       " 'DDI2',\n",
       " 'DDI2',\n",
       " 'RSC1A1',\n",
       " 'PLEKHM2',\n",
       " 'PLEKHM2',\n",
       " 'SLC25A34',\n",
       " 'TMEM82',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'FBLIM1',\n",
       " 'SPEN',\n",
       " 'SPEN',\n",
       " 'SPEN',\n",
       " 'ZBTB17',\n",
       " 'ZBTB17',\n",
       " 'ZBTB17',\n",
       " 'ZBTB17',\n",
       " 'ZBTB17',\n",
       " 'ZBTB17',\n",
       " 'C1orf64',\n",
       " 'HSPB7',\n",
       " 'HSPB7',\n",
       " 'HSPB7',\n",
       " 'HSPB7',\n",
       " 'HSPB7',\n",
       " 'HSPB7',\n",
       " 'HSPB7',\n",
       " 'CLCNKA',\n",
       " 'CLCNKA',\n",
       " 'CLCNKA',\n",
       " 'CLCNKA',\n",
       " 'CLCNKB',\n",
       " 'CLCNKB',\n",
       " 'CLCNKB',\n",
       " 'FAM131C',\n",
       " 'EPHA2',\n",
       " 'ARHGEF19',\n",
       " 'ARHGEF19',\n",
       " 'ARHGEF19',\n",
       " 'C1orf134',\n",
       " 'RSG1',\n",
       " 'RSG1',\n",
       " 'FBXO42',\n",
       " 'FBXO42',\n",
       " 'FBXO42',\n",
       " 'SZRD1',\n",
       " 'SZRD1',\n",
       " 'SZRD1',\n",
       " 'SZRD1',\n",
       " 'SZRD1',\n",
       " 'SZRD1',\n",
       " 'SPATA21',\n",
       " 'SPATA21',\n",
       " 'SPATA21',\n",
       " 'SPATA21',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'NECAP2',\n",
       " 'AL355149',\n",
       " 'NBPF1',\n",
       " 'NBPF1',\n",
       " 'NBPF1',\n",
       " 'NBPF1',\n",
       " 'NBPF1',\n",
       " 'AL137798',\n",
       " 'AL021920',\n",
       " 'CROCC',\n",
       " 'CROCC',\n",
       " 'MFAP2',\n",
       " 'MFAP2',\n",
       " 'MFAP2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'ATP13A2',\n",
       " 'SDHB',\n",
       " 'PADI2',\n",
       " 'PADI2',\n",
       " 'PADI2',\n",
       " 'PADI1',\n",
       " 'PADI1',\n",
       " 'PADI1',\n",
       " 'PADI1',\n",
       " 'PADI3',\n",
       " 'PADI4',\n",
       " ...]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[n.split(\".\")[0].split(\"-\")[0] for n in gene_names[4]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>918</td>\n",
       "      <td>CDS:1-918</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL627309.1-201</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>2661</td>\n",
       "      <td>UTR5:1-70</td>\n",
       "      <td>CDS:71-850</td>\n",
       "      <td>UTR3:851-2661</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR4F29-001</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>995</td>\n",
       "      <td>UTR5:1-19</td>\n",
       "      <td>CDS:20-958</td>\n",
       "      <td>UTR3:959-995</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR4F16-001</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>995</td>\n",
       "      <td>UTR5:1-19</td>\n",
       "      <td>CDS:20-958</td>\n",
       "      <td>UTR3:959-995</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL669831.1-201</td>\n",
       "      <td>AL669831.1</td>\n",
       "      <td>129</td>\n",
       "      <td>CDS:1-129</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AL645608.2-201</td>\n",
       "      <td>AL645608.2</td>\n",
       "      <td>57</td>\n",
       "      <td>CDS:1-57</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMD11-011</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>626</td>\n",
       "      <td>UTR5:1-89</td>\n",
       "      <td>CDS:90-626</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMD11-003</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>387</td>\n",
       "      <td>UTR5:1-60</td>\n",
       "      <td>CDS:61-387</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMD11-010</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2551</td>\n",
       "      <td>UTR5:1-83</td>\n",
       "      <td>CDS:84-2129</td>\n",
       "      <td>UTR3:2130-2551</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMD11-001</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>2191</td>\n",
       "      <td>CDS:1-1769</td>\n",
       "      <td>UTR3:1770-2191</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAMD11-004</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1731</td>\n",
       "      <td>CDS:1-1625</td>\n",
       "      <td>UTR3:1626-1731</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AL645608.1-201</td>\n",
       "      <td>AL645608.1</td>\n",
       "      <td>336</td>\n",
       "      <td>CDS:1-336</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NOC2L-001</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>2790</td>\n",
       "      <td>UTR5:1-50</td>\n",
       "      <td>CDS:51-2300</td>\n",
       "      <td>UTR3:2301-2790</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>KLHL17-001</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>2560</td>\n",
       "      <td>UTR5:1-107</td>\n",
       "      <td>CDS:108-2036</td>\n",
       "      <td>UTR3:2037-2560</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>KLHL17-004</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>913</td>\n",
       "      <td>CDS:1-463</td>\n",
       "      <td>UTR3:464-913</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PLEKHN1-001</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>2404</td>\n",
       "      <td>UTR5:1-35</td>\n",
       "      <td>CDS:36-1871</td>\n",
       "      <td>UTR3:1872-2404</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PLEKHN1-005</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>2455</td>\n",
       "      <td>UTR5:1-30</td>\n",
       "      <td>CDS:31-2022</td>\n",
       "      <td>UTR3:2023-2455</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PLEKHN1-004</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>2194</td>\n",
       "      <td>UTR5:1-30</td>\n",
       "      <td>CDS:31-1761</td>\n",
       "      <td>UTR3:1762-2194</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PLEKHN1-003</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>698</td>\n",
       "      <td>CDS:1-531</td>\n",
       "      <td>UTR3:532-698</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>C1orf170-201</td>\n",
       "      <td>C1orf170</td>\n",
       "      <td>3064</td>\n",
       "      <td>CDS:1-2091</td>\n",
       "      <td>UTR3:2092-3064</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                4           5     6           7               8    \n",
       "0        OR4F5-001       OR4F5   918   CDS:1-918                  \\\n",
       "1   AL627309.1-201  AL627309.1  2661   UTR5:1-70      CDS:71-850   \n",
       "2       OR4F29-001      OR4F29   995   UTR5:1-19      CDS:20-958   \n",
       "3       OR4F16-001      OR4F16   995   UTR5:1-19      CDS:20-958   \n",
       "4   AL669831.1-201  AL669831.1   129   CDS:1-129                   \n",
       "5   AL645608.2-201  AL645608.2    57    CDS:1-57                   \n",
       "6       SAMD11-011      SAMD11   626   UTR5:1-89      CDS:90-626   \n",
       "7       SAMD11-003      SAMD11   387   UTR5:1-60      CDS:61-387   \n",
       "8       SAMD11-010      SAMD11  2551   UTR5:1-83     CDS:84-2129   \n",
       "9       SAMD11-001      SAMD11  2191  CDS:1-1769  UTR3:1770-2191   \n",
       "10      SAMD11-004      SAMD11  1731  CDS:1-1625  UTR3:1626-1731   \n",
       "11  AL645608.1-201  AL645608.1   336   CDS:1-336                   \n",
       "12       NOC2L-001       NOC2L  2790   UTR5:1-50     CDS:51-2300   \n",
       "13      KLHL17-001      KLHL17  2560  UTR5:1-107    CDS:108-2036   \n",
       "14      KLHL17-004      KLHL17   913   CDS:1-463    UTR3:464-913   \n",
       "15     PLEKHN1-001     PLEKHN1  2404   UTR5:1-35     CDS:36-1871   \n",
       "16     PLEKHN1-005     PLEKHN1  2455   UTR5:1-30     CDS:31-2022   \n",
       "17     PLEKHN1-004     PLEKHN1  2194   UTR5:1-30     CDS:31-1761   \n",
       "18     PLEKHN1-003     PLEKHN1   698   CDS:1-531    UTR3:532-698   \n",
       "19    C1orf170-201    C1orf170  3064  CDS:1-2091  UTR3:2092-3064   \n",
       "\n",
       "                9     10  \n",
       "0             None  None  \n",
       "1    UTR3:851-2661        \n",
       "2     UTR3:959-995        \n",
       "3     UTR3:959-995        \n",
       "4             None  None  \n",
       "5             None  None  \n",
       "6                   None  \n",
       "7                   None  \n",
       "8   UTR3:2130-2551        \n",
       "9                   None  \n",
       "10                  None  \n",
       "11            None  None  \n",
       "12  UTR3:2301-2790        \n",
       "13  UTR3:2037-2560        \n",
       "14                  None  \n",
       "15  UTR3:1872-2404        \n",
       "16  UTR3:2023-2455        \n",
       "17  UTR3:1762-2194        \n",
       "18                  None  \n",
       "19                  None  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names.iloc[0:20,4:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      0                  1                     2    \n",
      "0      ENST00000335137.3  ENSG00000186092.4  OTTHUMG00000001094.1  \\\n",
      "1      ENST00000423372.3  ENSG00000237683.5                     -   \n",
      "2      ENST00000426406.1  ENSG00000235249.1  OTTHUMG00000002860.1   \n",
      "3      ENST00000332831.2  ENSG00000185097.2  OTTHUMG00000002581.1   \n",
      "4      ENST00000599533.1  ENSG00000269831.1                     -   \n",
      "...                  ...                ...                   ...   \n",
      "95304  ENST00000361335.1  ENSG00000212907.2                     -   \n",
      "95305  ENST00000361381.2  ENSG00000198886.2                     -   \n",
      "95306  ENST00000361567.2  ENSG00000198786.2                     -   \n",
      "95307  ENST00000361681.2  ENSG00000198695.2                     -   \n",
      "95308  ENST00000361789.2  ENSG00000198727.2                     -   \n",
      "\n",
      "                         3               4           5     6           7    \n",
      "0      OTTHUMT00000003223.1       OR4F5-001       OR4F5   918   CDS:1-918  \\\n",
      "1                         -  AL627309.1-201  AL627309.1  2661   UTR5:1-70   \n",
      "2      OTTHUMT00000007999.1      OR4F29-001      OR4F29   995   UTR5:1-19   \n",
      "3      OTTHUMT00000007334.1      OR4F16-001      OR4F16   995   UTR5:1-19   \n",
      "4                         -  AL669831.1-201  AL669831.1   129   CDS:1-129   \n",
      "...                     ...             ...         ...   ...         ...   \n",
      "95304                     -     MT-ND4L-201     MT-ND4L   297   CDS:1-297   \n",
      "95305                     -      MT-ND4-201      MT-ND4  1378  CDS:1-1378   \n",
      "95306                     -      MT-ND5-201      MT-ND5  1812  CDS:1-1812   \n",
      "95307                     -      MT-ND6-201      MT-ND6   525   CDS:1-525   \n",
      "95308                     -      MT-CYB-201      MT-CYB  1141  CDS:1-1141   \n",
      "\n",
      "               8              9     10  \n",
      "0                           None  None  \n",
      "1      CDS:71-850  UTR3:851-2661        \n",
      "2      CDS:20-958   UTR3:959-995        \n",
      "3      CDS:20-958   UTR3:959-995        \n",
      "4                           None  None  \n",
      "...           ...            ...   ...  \n",
      "95304                       None  None  \n",
      "95305                       None  None  \n",
      "95306                       None  None  \n",
      "95307                       None  None  \n",
      "95308                       None  None  \n",
      "\n",
      "[95309 rows x 11 columns]\n"
     ]
    }
   ],
   "source": [
    "print(gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
